dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs

```diff
@@ -190,46 +180,48 @@ internal KernelFunction(string name, string? pluginName, string description, IRe
             // Invoke the function.
             functionResult = await this.InvokeCoreAsync(kernel, arguments, cancellationToken).ConfigureAwait(false);
 
-            // Invoke the post-invocation event handler. If it requests cancellation, throw.
-#pragma warning disable CS0618 // Events are deprecated
-            var invokedEventArgs = kernel.OnFunctionInvoked(this, arguments, functionResult);
-#pragma warning restore CS0618 // Events are deprecated
-
-            // Invoke the post-invocation filter. If it requests cancellation, throw.
+            // Invoke the post-invocation filter.
             var invokedContext = kernel.OnFunctionInvokedFilter(arguments, functionResult);
 
-            if (invokedEventArgs is not null)
-            {
-                // Apply any changes from the event handlers to final result.
-                functionResult = new FunctionResult(this, invokedEventArgs.ResultValue, functionResult.Culture, invokedEventArgs.Metadata ?? functionResult.Metadata);
-            }
-
             if (invokedContext is not null)
             {
                 // Apply any changes from the function filters to final result.
                 functionResult = new FunctionResult(this, invokedContext.ResultValue, functionResult.Culture, invokedContext.Metadata ?? functionResult.Metadata);
             }
 
-            if (invokedEventArgs?.Cancel is true)
+            logger.LogFunctionInvokedSuccess(this.Name);
+            logger.LogFunctionResultValue(functionResult.Value);
+
+            return functionResult;
+        }
+        catch (Exception ex)
+        {
+            functionResult = new(this);
+            var exception = HandleException(ex, logger, activity, this, kernel, arguments, functionResult, ref tags);
+
+            // Invoke post-invocation filter with exception.
+            var invokedContext = kernel.OnFunctionInvokedFilter(arguments, functionResult, exception);
+
+            // No registered filters, throw original exception.
+            if (invokedContext is null)
             {
-                throw new OperationCanceledException($"A {nameof(Kernel)}.{nameof(Kernel.FunctionInvoked)} event handler requested cancellation after function invocation.");
+                throw exception;
```

Comment:
We receive `exception` instance from method `HandleException`. In this method, we pass original exception. We either return it or in case if execution was cancelled, we wrap it into `KernelFunctionCanceledException`. I think in both cases stack trace should be preserved. Let me know if I miss something.